### PR TITLE
Fix hardcoded ROCm paths in `Caffe2Targets.cmake`

### LINKED ISSUE #131701: `/opt/rocm/lib/libamdhip64.so` is hardcoded in `Caffe2Targets.cmake` in ROCm wheels
### 🐛 Describe the bug

`/opt/rocm` is hardcoded in `Caffe2Targets.cmake` in ROCm wheels:

```sh
wget https://download.pytorch.org/whl/rocm6.1/torch-2.4.0%2Brocm6.1-cp312-cp312-linux_x86_64.whl
unzip torch-2.4.0+rocm6.1-cp312-cp312-linux_x86_64.whl
cd torch/share/cmake/Caffe2
```

You can find below in `Caffe2Targets.cmake`:

```cmake
set_target_properties(c10_hip PROPERTIES
  INTERFACE_INCLUDE_DIRECTORIES "${_IMPORT_PREFIX}/include"
  INTERFACE_LINK_LIBRARIES "c10;/opt/rocm/lib/libamdhip64.so"                                                                                                                                                                                                                                                                 
)
```

```cmake
set_target_properties(torch_hip PROPERTIES
  INTERFACE_COMPILE_DEFINITIONS "USE_C10D_NCCL"
  INTERFACE_COMPILE_OPTIONS "-fPIC;-D__HIP_PLATFORM_AMD__=1;-DCUDA_HAS_FP16=1;-DUSE_ROCM;-D__HIP_NO_HALF_OPERATORS__=1;-D__HIP_NO_HALF_CONVERSIONS__=1;-DTORCH_HIP_VERSION=601;-Wno-shift-count-negative;-Wno-shift-count-overflow;-Wno-duplicate-decl-specifier;-DCAFFE2_USE_MIOPEN;-DTHRUST_DEVICE_SYSTEM=THRUST_DEVICE_SYSTEM_HIP;-std=c++17;-DHIPBLAS_V2;-DHIP_NEW_TYPE_ENUMS"
  INTERFACE_INCLUDE_DIRECTORIES "${_IMPORT_PREFIX}/include"
  INTERFACE_LINK_LIBRARIES "c10_hip;torch_cpu_library;/opt/rocm/lib/libamdhip64.so;MIOpen;/opt/rocm/lib/libhiprtc.so;/opt/rocm/lib/libroctx64.so;roc::hipblaslt;roc::hipblas;hip::hipfft;hip::hiprand;roc::hipsparse;roc::hipsolver"
)
```

In some HPCs, the ROCm may not be installed in `/opt/rocm`, but in other directories such as `/opt/rocm-6.0.0`. When a third-party package uses `find_package(Torch)` and links with PyTorch libraries (which is suggested by [the tutorial](https://pytorch.org/tutorials/advanced/torch_script_custom_ops.html#building-with-cmake)), it may introduce an invalid link library.

@Slaedr originally posted this problem in https://github.com/deepmodeling/deepmd-kit/issues/4008 (our package that links with PyTorch libraries).

I noticed there was a similar issue in other PyTorch modules https://github.com/pytorch/kineto/issues/838.

### Versions

Not related to the runtime environment

cc @ezyang @gchanan @zou3519 @kadeng @msaroufim @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @dllehr-amd @jataylo @hongxiayang

### PR DESCRIPTION
**Pull request by naromero77amd**
Fixes #131701

Use CMake imported targets more consistently to eliminate hardcode paths.

Here is the new relevant sections of Caffe2Targets.cmake:
```
set_target_properties(c10_hip PROPERTIES
  INTERFACE_INCLUDE_DIRECTORIES "${_IMPORT_PREFIX}/include"
  INTERFACE_LINK_LIBRARIES "c10;hip::amdhip64"
)
```

```
set_target_properties(torch_hip PROPERTIES
  INTERFACE_COMPILE_DEFINITIONS "USE_C10D_NCCL"
  INTERFACE_COMPILE_OPTIONS "-fPIC;-D__HIP_PLATFORM_AMD__=1;-DCUDA_HAS_FP16=1;-DUSE_ROCM;-D__HIP_NO_HALF_OPERATORS__=1;-D__HIP_NO_HALF_CONVERSIONS__=1;-DTORCH_HIP_VERSION=602;-Wno-shift-count-negative;-Wno-shift-count-overflow;-Wno-duplicate-decl-specifier;-DCAFFE2_USE_MIOPEN;-DTHRUST_DEVICE_SYSTEM=THRUST_DEVICE_SYSTEM_HIP;-std=c++17;-DHIPBLAS_V2;-DHIP_NEW_TYPE_ENUMS"
  INTERFACE_INCLUDE_DIRECTORIES "${_IMPORT_PREFIX}/include"
  INTERFACE_LINK_LIBRARIES "c10_hip;torch_cpu_library;hip::amdhip64;MIOpen;hiprtc::hiprtc;roc::hipblaslt;roc::hipblas;hip::hipfft;hip::hiprand;roc::hipsparse;roc::hipsolver"
)
```

HIPCUB dependency was not actually used; which is why it is removed here as the imported target had undesirable side effects.

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @dllehr-amd @jataylo @hongxiayang

### CODE REVIEW

**Review by jeffdaily — APPROVED** [review]
sure why not.  if CI is green and this fixes the issue, good to go.

**Review by Skylion007 — APPROVED** [review]

File: c10/hip/CMakeLists.txt [hunk 1]
```diff
@@ -48,9 +48,7 @@ if(NOT BUILD_LIBTORCHLESS)
   endif()
 
   # ---[ Dependency of c10_hip
-  target_link_libraries(c10_hip PUBLIC c10)
-
-  target_link_libraries(c10_hip PUBLIC ${PYTORCH_HIP_LIBRARIES})
+  target_link_libraries(c10_hip PUBLIC ${C10_LIB} hip::amdhip64)
```
Comment (jithunnair-amd) [inline c10/hip/CMakeLists.txt:51]: @naromero77amd Why the change from `c10` to `${C10_LIB}`?

File: cmake/Dependencies.cmake [hunk 2]
```diff
@@ -1093,8 +1093,8 @@ if(USE_ROCM)
     hip_include_directories(${Caffe2_HIP_INCLUDE})
 
     set(Caffe2_PUBLIC_HIP_DEPENDENCY_LIBS
-      ${PYTORCH_HIP_LIBRARIES} ${PYTORCH_MIOPEN_LIBRARIES} ${hipcub_LIBRARIES} ${ROCM_HIPRTC_LIB} ${ROCM_ROCTX_LIB})
-    list(APPEND Caffe2_PUBLIC_HIP_DEPENDENCY_LIBS ${hipblaslt_LIBRARIES})
+      hip::amdhip64 MIOpen hiprtc::hiprtc) # libroctx will be linked in with MIOpen
```
Comment (jithunnair-amd) [inline cmake/Dependencies.cmake:1096]: @naromero77amd I'd also mention roctx as a direct dependency, even if it's pulled in as a dependency of MIOpen, because PyTorch does depend on roctx, independent of its dependency on MIOpen? Also, if the only usage of variables like `${PYTORCH_MIOPEN_LIBRARIES` was here, can we clean up the code that defines them in LoadHip.cmake? (another one is `PYTORCH_RCCL_LIBRARIES`, which is used in [rccl.cmake](https://github.com/search?q=repo%3Apytorch%2Fpytorch%20PYTORCH_RCCL_LIBRARIES&type=code))

**Review by jithunnair-amd — CHANGES_REQUESTED** [review]

File: c10/hip/CMakeLists.txt (hunk 1 above)
Comment (naromero77amd) [reply c10/hip/CMakeLists.txt:51]: This is for consistency with what is done in [c10/cuda/CMakeLists.txt](https://github.com/pytorch/pytorch/blob/main/c10/cuda/CMakeLists.txt#L68)

**Review by atalman — APPROVED** [review]
lgtm

File: cmake/Dependencies.cmake (hunk 2 above)
Comment (jithunnair-amd) [reply cmake/Dependencies.cmake:1096]: Capturing offline discussion: this PR will limit itself to changes to address https://github.com/pytorch/pytorch/issues/131701, and a future PR will address further cleanup.

**Review by jithunnair-amd — APPROVED** [review]
